### PR TITLE
(LedgerStore) Implement new() for LedgerStoreOptions

### PR DESCRIPTION
**Pull request by yhchiang-sol**
#### Problem
As we start adding private fields to LedgerStoreOptions (e.g. in #24684), LedgerStoreOptions needs
`new()` in order to initialize private fields.

#### Summary of Changes
Implement `new()` for LedgerStoreOptions.

This PR depends on #24682 and #24761.

### CODE REVIEW

File: core/tests/ledger_cleanup.rs [hunk 1]
```diff
@@ -351,15 +351,14 @@ mod tests {
...
-                        ),
-                        ..LedgerColumnOptions::default()
-                    },
+                    column_options: LedgerColumnOptions::new(
+                        ShredStorageType::RocksFifo(BlockstoreRocksFifoOptions {
+                            shred_data_cf_size: config.shred_data_cf_size,
+                            ..BlockstoreRocksFifoOptions::default()
+                        }),
+                        BlockstoreCompressionType::default(),
+                        0,
```
Comment (behzadnouri) [inline core/tests/ledger_cleanup.rs:360]: For the next person reading this code, please add a comment what this zero is; as in
```rust
  0, // rocks_perf_sample_interval
```

File: ledger/src/blockstore.rs [hunk 2]
```diff
@@ -3978,12 +3981,13 @@ macro_rules! create_new_tmp_ledger_fifo_auto_delete {
...
-            $crate::blockstore_db::LedgerColumnOptions {
-                shred_storage_type: $crate::blockstore_db::ShredStorageType::RocksFifo(
+            $crate::blockstore_db::LedgerColumnOptions::new(
+                $crate::blockstore_db::ShredStorageType::RocksFifo(
                     $crate::blockstore_db::BlockstoreRocksFifoOptions::default(),
                 ),
-                ..$crate::blockstore_db::LedgerColumnOptions::default()
-            },
+                BlockstoreCompressionType::default(),
+                0,
```
Comment (behzadnouri) [inline ledger/src/blockstore.rs:3989]: similarly here
```rust
  0, // rocks_perf_sample_interval
```

File: validator/src/main.rs [hunk 3]
```diff
@@ -2706,44 +2706,45 @@ pub fn main() {
...
-                "none" => BlockstoreCompressionType::None,
-                "snappy" => BlockstoreCompressionType::Snappy,
-                "lz4" => BlockstoreCompressionType::Lz4,
-                "zlib" => BlockstoreCompressionType::Zlib,
-                _ => panic!(
-                    "Unsupported ledger_compression: {}",
-                    ledger_compression_string
-                ),
-            },
+    let rocks_compression_type = match matches.value_of("rocksdb_ledger_compression") {
```
Comment (behzadnouri) [inline validator/src/main.rs:2709]: Might be better to move all these code and below to a helper function in `validator/src/main.rs`;
Something like:
```rust
fn make_ledger_column_options(matches: &ArgMatches<'_>) -> LedgerColumnOptions {
   ...
}
```
and then in `fn main()` just do
```rust
validator_config.ledger_column_options = make_ledger_column_options(&matches);
```

**Comment by yhchiang-sol** [comment]
The `new` function approach is more suitable than the `pub(crate)` solution, as LedgerColumnOptions will also later include the struct that we locally collect the aggregated metrics and report periodically.

If we really hit the situation where multiple arguments might be confusing, then a more robust way would be creating a new struct called `LedgerColumnOptionsBuilder` and implementing a `set` function for every public field.  At this moment `new` function approach should be acceptable IMO.

**Review by steviez — COMMENTED** [review]
> as LedgerColumnOptions will also later include the struct that we locally collect the aggregated metrics and report periodically.

I could be misunderstanding the intention, but this new struct sounds like it wouldn't really be an option, so perhaps it'd be better suited as an additional struct within `LedgerColumn` ? If so, I'm leaning towards agreeing with Behzad with the struct syntax being self-documenting / sufficient. Throughout the codebase, it seems we stick with the struct syntax for the config only structs (such as `BlockstoreOptions`)

**Comment by yhchiang-sol** [comment]
> I could be misunderstanding the intention, but this new struct sounds like it wouldn't really be an option, so perhaps it'd be better suited as an additional struct within LedgerColumn ? If so, I'm leaning towards agreeing with Behzad with the struct syntax being self-documenting / sufficient. Throughout the codebase, it seems we stick with the struct syntax for the config only structs (such as BlockstoreOptions)

I think having a separate struct within `LedgerColumn` would be the best for readability and the code structure.  Originally I am thinking about passing one-less `Arc&` argument by having the struct within LedgerColumnOptions.

If we feel having a separate struct within LedgerColumn sounds a better solution, then these two internal counters can also be moved to that struct so that we can keep our LedgerColumnOptions cleaner.

**Comment by steviez** [comment]
> I think having a separate struct within `LedgerColumn` would be the best for readability and the code structure. Originally I am thinking about passing one-less `Arc&` argument by having the struct within LedgerColumnOptions.
> 
> If we feel having a separate struct within LedgerColumn sounds a better solution, then these two internal counters can also be moved to that struct so that we can keep our LedgerColumnOptions cleaner.

We could keep `LedgerColumnOptions` as the struct that gets passed down from validator (validator need not worry about lower level internals). Then, `LedgerColumnOptions` (or just the fields) could get thrown into a `LedgerColumnMetrics` struct (maybe different name if we have more stuff that we want to throw in) that `LedgerColumn` would keep an `Arc` to.

Regardless, should we move any further conversation on this to https://github.com/solana-labs/solana/pull/24684. For the scope of this PR, I think it sounds like we're leaning towards not adding a `new()` function ?

**Comment by yhchiang-sol** [comment]
Agree.  Let's abandon this one and continue the discussion on #24684.

**Pull request closed without merging**